src/main/java/pl/javastart/streamstask/StreamsTask.java

```diff
@@ -32,28 +31,70 @@ private void run() {
+        Map<Long, List<Expense>> expensesByUserIdVersion2 = groupExpensesByUserIdVersion2(users, expenses);
+        Map<User, List<Expense>> expensesByUserVersion1 = groupExpensesByUserVersion1(users, expenses);
+        Map<User, List<Expense>> expensesByUserVersion2 = groupExpensesByUserVersion2(users, expenses);
     }
 
     // metoda powinna zwracać listę kobiet (sprawdzając, czy imię kończy się na "a")
     Collection<User> findWomen(Collection<User> users) {
-        throw new RuntimeException("Not implemented");
+        return users.stream()
+                .filter(u -> u.getName().endsWith("a"))
+                .collect(Collectors.toList());
     }
 
     // metoda powinna zwracać średni wiek mężczyzn (sprawdzając, czy imię nie kończy się na "a")
     Double averageMenAge(Collection<User> users) {
-        throw new RuntimeException("Not implemented");
+        return users.stream()
+                .filter(u -> !u.getName().endsWith("a"))
+                .mapToInt(User::getAge)
+                .average()
+                .getAsDouble();
     }
 
     // metoda powinna zwracać wydatki zgrupowane po ID użytkownika
-    Map<Long, List<Expense>> groupExpensesByUserId(Collection<User> users, List<Expense> expenses) {
-        throw new RuntimeException("Not implemented");
+    // Collection<User> users wydał mi się niepotrzebny jako argument, dlatego go usunąłem
+    Map<Long, List<Expense>> groupExpensesByUserIdVersion1(List<Expense> expenses) {
+        return expenses.stream().collect(Collectors.groupingBy(Expense::getUserId));
+    }
+
+    // Druga wersja powyższej metody, tym razem zestawiająca w oryginalnej kolejności (dzięki LinkedHashMap)
+    // wszystkich userów, nie tylko tych z jakimiś wydatkami
+    Map<Long, List<Expense>> groupExpensesByUserIdVersion2(Collection<User> users, List<Expense> expenses) {
+        return users.stream()
+                .collect(Collectors.toMap(
+                        User::getId,
+                        user -> expenses.stream()
+                                .filter(expense -> expense.getUserId() == user.getId())
+                                .collect(Collectors.toList()),
+                        (x, y) -> y,
+                        LinkedHashMap::new
+                ));
     }
 
     // metoda powinna zwracać wydatki zgrupowane po użytkowniku
     // podobne do poprzedniego, ale trochę trudniejsze
-    Map<User, List<Expense>> groupExpensesByUser(Collection<User> users, List<Expense> expenses) {
-        throw new RuntimeException("Not implemented");
+    Map<User, List<Expense>> groupExpensesByUserVersion1(Collection<User> users, List<Expense> expenses) {
+        return expenses.stream()
+                .collect(Collectors.groupingBy(expense -> users.stream()
+                        .filter(u -> u.getId() == expense.getUserId())
+                        .findFirst()
+                        .orElseThrow()
+                ));
```

Comment:
Lepiej byłoby stworzyć sobie mapę userId->User, byłoby to dużo wydajniejsze, niż tworzenie strumienia i przeszukiwanie kolekcji Userów dla każdego wydatku